common/src/main/java/com/wynntils/core/storage/StorageManager.java

```diff
@@ -0,0 +1,119 @@
+ */
+package com.wynntils.core.storage;
+
+import com.google.gson.JsonElement;
+import com.google.gson.JsonObject;
+import com.wynntils.core.WynntilsMod;
+import com.wynntils.core.components.Manager;
+import com.wynntils.core.components.Managers;
+import com.wynntils.core.json.JsonManager;
+import com.wynntils.core.mod.event.WynncraftConnectionEvent;
+import com.wynntils.utils.mc.McUtils;
+import java.io.File;
+import java.lang.reflect.Field;
+import java.lang.reflect.Type;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.TreeMap;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import net.minecraftforge.eventbus.api.SubscribeEvent;
+import org.apache.commons.lang3.reflect.FieldUtils;
+
+public final class StorageManager extends Manager {
+    private static final long SAVE_INTERVAL = 10_000;
+
+    private static final File STORAGE_DIR = WynntilsMod.getModStorageDir("storage");
+    private static final String FILE_SUFFIX = ".data.json";
+    private final File userStorageFile;
+
+    private final ScheduledExecutorService executor = Executors.newScheduledThreadPool(1);
+    private final Map<String, Storage> storages = new TreeMap<>();
+    private final Map<Storage, Type> storageTypes = new HashMap<>();
+
+    private long lastPersisted;
+    private boolean scheduledPersist;
+
+    public StorageManager(JsonManager jsonManager) {
+        super(List.of(jsonManager));
+        userStorageFile = new File(STORAGE_DIR, McUtils.mc().getUser().getUuid() + FILE_SUFFIX);
+    }
+
+    public void registerStorageable(Storageable storageable) {
+        String baseName = storageable.getStorageJsonName();
+
+        List<Field> fields = FieldUtils.getAllFieldsList(storageable.getClass());
+        List<Field> storageFields =
+                fields.stream().filter(f -> f.getType().equals(Storage.class)).toList();
+
+        for (Field storageField : storageFields) {
+            try {
+                Storage storage = (Storage) FieldUtils.readField(storageField, storageable, true);
+                String jsonName = baseName + "." + storageField.getName();
+                storages.put(jsonName, storage);
+
+                Type typeOverride = Managers.Json.findFieldTypeOverride(storageable, storageField);
+                storageTypes.put(
+                        storage,
+                        typeOverride != null ? typeOverride : storage.get().getClass());
+            } catch (IllegalAccessException e) {
+                throw new RuntimeException(e);
+            }
+        }
+    }
+
+    public void restorePersisted() {
+        readFromJson();
+    }
+
+    @SubscribeEvent
+    public void onWynncraftDisconnect(WynncraftConnectionEvent.Disconnected event) {
+        // Always save when disconnecting
+        writeToJson();
+    }
+
+    void persist(Storage<?> storage) {
```

Comment:
So basically we save our storage every 10 secs, if there needs to be save (so a storage was changed)?

Reply:
Yep. Or to be more precise, we never store more often than every tenth second. If no change has been made for more than 10 seconds, the data is stored immediately (or, more or less immediately, it is done in a separate thread, to not cause lag spikes on the rendering thread). But if data has been saved less than ten seconds ago, we schedule a future write 10 seconds after the last.

So if we write value A, and then 8 seconds later get a request to write B, we will wait 2 second. If a request to write value C comes in after an additional 1 seconds, it will see that a write is scheduled and is happy with that, so what we get is:
T = 0: Write A
T = 10: Write B and C